### ISSUE TITLE
Feature Request: Use alternate assignments to infer type of empty list or dictionary

### ISSUE THREAD

**Issue by ipkiss42** [microsoft/pyright#1670]
**Is your feature request related to a problem? Please describe.**
Pyright does not seem to support type inference. For example, in line 4 of the following code `var` has an unknown type:

```python
if 1 == 2:
    var = [123]
else:
    var = []
```

**Describe the solution you'd like**
Consider supporting type inference in the [same way as mypy](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html). In particular: "Mypy considers the initial assignment as the definition of a variable.".

**Additional context**
An obvious workaround is to add a type hint for `var`:

```python
var: List[int]
if 1 == 2:
    var = [123]
else:
    var = []
```

However, mypy's type inference provides a nice benefit in that it avoids explicit typing of most local variables. In turn, this helps avoid redundant typing (DRY principle).

For example, with the following code, changing the type of the list (e.g. to `List[str]`) means changing only the function signature (and the code generating the list, of course), but with an additional type hint then the type hint would also have to be changed.

```python
def foo() -> List[int]:
    if 1 == 2:
        var = [123]
    else:
        var = []
    return var
```

**Comment by erictraut** [comment]
Pyright's approach to inference is intentionally different from mypy. Pyright's inference allows you to avoid explicit typing of most local variables and in more cases than mypy does.

As you point out, mypy considers the initial assignment as the definition of a variable, but this causes _many_ problems, not only for local variables but also for instance and class variables. If you read through the active bug reports logged against mypy, you'll find that many users find mypy's approach to be problematic.

Pyright's inference engine considers _all_ assignments of a variable rather than just the first assignment. Refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md) for how inference works in pyright.

Consider the following:
```python
if 1 == 2:
   var = 3
else:
   var = 5.6
```
This works fine in pyright but requires a type annotation in mypy.

The specific example you provided above is a rare example of where pyright requires a type annotation whereas mypy doesn't.

**Comment by ipkiss42** [comment]
Thanks for the link, the documentation is very well written! I agree that considering all the assignments is better than considering only the first one.

The only piece of information I am missing from the documentation is what happens to `Unknown`. According to the "Multi-Assignment Type Inference" section, in my original example `var` should have an inferred type of `Union[List[int], List[Unknown]]`.

So I think my question becomes: could `Union[List[SomeType], List[Unknown]]` be inferred as `List[SomeType]`? Or maybe via an intermediate step, inferred first as `List[Union[SomeType, Unknown]]`, and then as `List[SomeType]`?

Some of this inference would be a heuristic, and like any heuristic it might bring false positives. But when the union contains only one type other than `Unknown`, isn't it a reasonable heuristic?

BTW, in my last example, the type hint for the return type of `foo()` is an additional hint that `List[Unknown]` could/should be coalesced into a `List[int]`. I'm not sure if this is something that Pyright could leverage, though (it might be asking too much from the bidirectional type inference).

**Comment by erictraut** [comment]
Ignoring or coalescing an "Unknown" would be dangerous because an "Unknown" can come from many sources, including unannotated types that come from imported symbols. If you don't care about "Unknown" types, you can disable the "reportUnknownVariableType" check. The purpose of this check is to notify you when an "Unknown" type is assigned to a variable.

Let me explore whether it would be feasible to special-case the assignment of empty lists (`[]`) and empty dictionaries (`{}`) if there are other assignments that provide non-empty contents.

**Comment by erictraut** [comment]
I figured out a clever way to make this case work in the most common cases.

Here's [documentation](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#empty-list-and-dictionary-type-inference) for this new capability. It works with both local variables and instance variables. Additional examples can be found in the [unit test for the feature](https://github.com/microsoft/pyright/blob/master/packages/pyright-internal/src/tests/samples/emptyContainers1.py).

Thanks again for the suggestion.

**Comment by erictraut** [comment]
This is included in pyright 1.1.125, which I just published. It will also be included in the next release of pylance.

**Comment by ipkiss42** [comment]
This is great, thank you for this improvement and all your awesome work on Pyright!
